From 172e6f4bf18a8a306aabdd707f76869c462fa1f1 Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Mon, 18 Jan 2021 22:24:40 +0200
Subject: [PATCH] [SW-20810] sha256: use builtin bswap32

Use GCC builtin bswap32 routine instead of swapping macro. This is done
in order to reduce the image footprint when introducing strict-align GCC
flag which changes the ASM code generated by the swapping macro.

The builtin bswap32 forces the GCC to use REV instruction in ARM instead
of using the byte shifting as written in macro.

Before the introduction of strict-align flag, the implication on
footprint is minimal at best since the GCC optimization replaces the
byte shifting code with REV instruction. This optimization stops when
using the strict-align flag.

Since the data "fed" to bswap32 is 32bit aligned, it is safe to use.

Change-Id: Ic48422c1d385399e9195aa48e9b0d7113a6dab5f
Signed-off-by: Guy Nisan <gnisan@habana.ai>
(cherry picked from commit 1da87d1c3ecc2c3b8de64f49dc0676184ed62295)
---

diff --git a/common/sha256.c b/common/sha256.c
index b696d68..3677936 100644
--- a/common/sha256.c
+++ b/common/sha256.c
@@ -9,6 +9,8 @@
 #include <string.h>
 #include "sha256.h"
 
+#define USE_BSWAP32
+
 /*----------------------------------------------------------------*/
 
 const u8 sha256_der_prefix[SHA256_DER_LEN] = {
@@ -17,24 +19,13 @@
 	0x00, 0x04, 0x20
 };
 
-/*
- * 32-bit integer manipulation macros (big endian)
- */
+/* Using __builtin_bswap32 reduces the image footprint */
+
 #ifndef GET_UINT32_BE
-#define GET_UINT32_BE(n, b, i) {			\
-	(n) = ((unsigned long) (b)[(i)] << 24)		\
-	    | ((unsigned long) (b)[(i) + 1] << 16)	\
-	    | ((unsigned long) (b)[(i) + 2] <<  8)	\
-	    | ((unsigned long) (b)[(i) + 3]);		\
-}
+#define GET_UINT32_BE(n, b, i) { n = __builtin_bswap32(*(u32 *)&b[i]); }
 #endif
 #ifndef PUT_UINT32_BE
-#define PUT_UINT32_BE(n, b, i) {			\
-	(b)[(i)] = (unsigned char)((n) >> 24);		\
-	(b)[(i) + 1] = (unsigned char) ((n) >> 16);	\
-	(b)[(i) + 2] = (unsigned char) ((n) >>  8);	\
-	(b)[(i) + 3] = (unsigned char) ((n));		\
-}
+#define PUT_UINT32_BE(n, b, i) { *((u32 *)&b[i]) = __builtin_bswap32(n); }
 #endif
 
 void sha256_starts(struct sha256_context *ctx)
